File: 2018/day01/solution.hs
import qualified Data.Set as S
import Data.Function ((&))

toInt ('+':n) = read n
toInt n = read n

duplicates xs =
    dups xs S.empty
    where dups (y:ys) s | S.member y s = (y:dups ys s) | otherwise = dups ys (S.insert y s)

accum = scanl (+) 0

main = do
    input <- map toInt . lines <$> readFile "2018/day01/in.txt"
    print $ sum input
    print $ input & cycle & accum & duplicates & head

File: 2018/day01/solution.py
from itertools import accumulate, cycle

def parse_input():
    return (int(n) for n in open('2018/day01/in.txt', 'r').readlines())

# PART 1
print(sum(parse_input()))


# PART 2
def duplicates(xs):
    seen = set([0])
    for x in xs:
        if x in seen:
            yield x
        else:
            seen.add(x)

print(next(duplicates(accumulate(cycle(parse_input())))))